test/e2e/ocpbugs48050_test.go

```diff
@@ -0,0 +1,611 @@
+				admittedRoutes++
+			} else {
+				pendingRoutes = append(pendingRoutes, fmt.Sprintf("%s/%s", routeList.Items[i].Namespace, routeList.Items[i].Name))
+			}
+		}
+
+		totalRoutes := len(routeList.Items)
+		if progress != nil {
+			progress(admittedRoutes, totalRoutes, pendingRoutes)
+		}
+
+		if admittedRoutes == totalRoutes {
+			return true, nil
+		}
+
+		return false, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("not all routes were admitted in namespace %s: %v", namespace, err)
+	}
+
+	return &routeList, nil
+}
+
+func newPrometheusClient() (prometheusv1client.API, error) {
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Kubernetes configuration: %v", err)
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client: %v", err)
+	}
+
+	routeClient, err := routev1client.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create route client: %v", err)
+	}
+
+	prometheusClient, err := metrics.NewPrometheusClient(context.TODO(), kubeClient, routeClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Prometheus client: %v", err)
+	}
+
+	return prometheusClient, nil
+}
+
+func queryPrometheus(client prometheusv1client.API, query string, timeout time.Duration) (model.Value, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	result, warnings, err := client.Query(ctx, query, time.Now())
+	if err != nil {
+		return nil, fmt.Errorf("error querying Prometheus: %v", err)
+	}
+
+	if len(warnings) > 0 {
+		return nil, fmt.Errorf("warnings querying Prometheus: %v", warnings)
+	}
+
+	return result, nil
+}
+
+type podScrapeCounts map[string]float64
+
+func getTotalScrapesForPods(promClient prometheusv1client.API, timeout time.Duration) (podScrapeCounts, error) {
+	query := `sum by (pod) (haproxy_exporter_total_scrapes{pod=~"router-default-.*"})`
+
+	result, err := queryPrometheus(promClient, query, timeout)
+	if err != nil {
+		return nil, err
+		//t.Fatalf("Failed to query Prometheus for total scrapes: %v", err)
+	}
+
+	vector, ok := result.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("expected result type model.Vector, got %T", result)
+	}
+
+	scrapeCounts := make(map[string]float64)
+
+	for _, sample := range vector {
+		podName := string(sample.Metric["pod"])
+		scrapeCounts[podName] = float64(sample.Value)
+	}
+
+	return scrapeCounts, nil
+}
+
+func waitForDefaultRouterScrapesIncrement(promClient prometheusv1client.API, minNewScrapes int, timeout time.Duration, progress func(initialCounts, currentCounts podScrapeCounts, allIncremented bool)) error {
+	initialCounts, err := getTotalScrapesForPods(promClient, timeout)
+	if err != nil {
+		return fmt.Errorf("failed to get initial scrape counts: %v", err)
+	}
+
+	deadline := time.Now().Add(timeout)
+
+	for time.Now().Before(deadline) {
```

Comment:
nit: similar logic is often implemented using a ticker and a timer. [Example](https://github.com/openshift/aws-load-balancer-operator/blob/721cf7c520603efcc9f81ee7e218bc6c66fda575/main.go#L240-L255).

Reply:
Will it be materially different for this test case?

Reply:
Not much difference logic wise, just more go idiomatic and maybe shorter (quicker too? hard to say).